markhor_flippers/include/markhor_hw_interface_flippers.hpp

```diff
@@ -0,0 +1,108 @@
+#ifndef MARKHOR_HW_INTERFACE_H
+#define MARKHOR_HW_INTERFACE_H
+
+// ROS
+#include <ros/ros.h>
+#include <std_msgs/Float32.h>
+#include <std_srvs/Empty.h>
+
+// ros_control
+#include <controller_manager/controller_manager.h>
+#include <hardware_interface/joint_command_interface.h>
+#include <hardware_interface/joint_state_interface.h>
+#include <hardware_interface/robot_hw.h>
+
+// ostringstream
+#include <sstream>
+#include <string>
+
+// CTRE
+#include "Platform-linux-socket-can.h"
+#include "ctre/Phoenix.h"
+#include "ctre/phoenix/motorcontrol/can/TalonSRX.h"
+#include "ctre/phoenix/unmanaged/Unmanaged.h"
+#include "geometry_msgs/Twist.h"
+#include "std_msgs/String.h"
+
+// STD
+#include <chrono>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+#include <iostream>
+#include <fstream>
+
+class MarkhorHWInterfaceFlippers : public hardware_interface::RobotHW
+{
+public:
+  MarkhorHWInterfaceFlippers();
+  void write();
+  void read();
+
+  ros::NodeHandle nh;
+  int num_joints = 0;
+
+  int currentLimit = 30;
+  
+private:
+  void setupRosControl();
+  void setupCtreDrive();
+  bool hasResetOccurred();
+  void printDriveInfo(std::unique_ptr<TalonSRX>& drive);
+  
+  void saveDrivePosition();
+  void writeDrivePositionToFile(std::string config_file_name);
+
+  void loadDrivePosition();
+  void readDrivePositionFromFile(std::string config_file_name, std::fstream& config_file);
+  void parseDrivePosition(std::string line);
+  void applyDrivePosition(std::unique_ptr<TalonSRX>& drive, float drive_position);
+
+  int getEncoderPosition(std::unique_ptr<TalonSRX>& drive);
+
+  const int kTimeoutMs = 30;
+  
+  int drive_fl_id, drive_fr_id, drive_rl_id, drive_rr_id = 0;
+
+  hardware_interface::JointStateInterface joint_state_interface_;
+  hardware_interface::PositionJointInterface position_joint_interface_;
+  std::vector<std::string> joint_names_;
+
+  std::vector<double> joint_position_;
+  std::vector<double> joint_velocity_;
+  std::vector<double> joint_effort_;
+
+  std::vector<double> joint_position_command_;
+  std::vector<double> joint_velocity_command_;
+  std::vector<double> joint_effort_command_;
+
+  std::unique_ptr<TalonSRX> front_left_drive;
```

Comment:
So, there's a lot of repeated logic that's pretty much identical for all the drives. For examples, things like printing the state of the drive or initializing the drive. All drives also have a upper/lower limit. I think this shows there's a missing abstraction that wraps the drive in a structure containing the pointer to the drive and the associated limits. Once you have that structure it would also be much easier to juste extract shared code like initializing and printing the state.

Reply:
I've created an issue for it